File: mobile/lib/features/announcements/domain/entities/announcement.dart
/// Một thông báo (kỳ thi JLPT / khoá MBA / chủ đề khác) do backend
/// KatoAssistant crawl từ nguồn GỐC chính thức, đã diff & (tùy chọn) xác thực.
///
/// `sourceUrl` + `sourceDomain` luôn là nguồn chính thức để người dùng tự kiểm
/// chứng (chống tin giả). `contentHash` ổn định theo nội dung → khoá chống báo
/// lại ở phía app.
class Announcement {
  final int id;
  final String topic; // jlpt | mba | custom
  final String title;
  final String summary;
  final String sourceUrl;
  final String sourceDomain;
  final DateTime firstSeenAt;
  final String contentHash;
  final bool verified;
  final double score;

  const Announcement({
    required this.id,
    required this.topic,
    required this.title,
    required this.summary,
    required this.sourceUrl,
    required this.sourceDomain,
    required this.firstSeenAt,
    required this.contentHash,
    required this.verified,
    required this.score,
  });

  factory Announcement.fromJson(Map<String, dynamic> json) {
    return Announcement(
      id: (json['id'] as num).toInt(),
      topic: json['topic'] as String? ?? 'custom',
      title: json['title'] as String? ?? '',
      summary: json['summary'] as String? ?? '',
      sourceUrl: json['source_url'] as String? ?? '',
      sourceDomain: json['source_domain'] as String? ?? '',
      firstSeenAt: DateTime.tryParse(json['first_seen_at'] as String? ?? '')
              ?.toLocal() ??
          DateTime.now(),
      contentHash: json['content_hash'] as String? ?? '',
      verified: json['verified'] as bool? ?? false,
      score: (json['score'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

File: backend/app/models/announcement.py
"""Bảng announcements — mỗi thông báo (đã diff & xác thực) từ nguồn chính thức."""

from datetime import datetime

from sqlalchemy import Boolean, DateTime, Float, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class Announcement(Base):
    __tablename__ = "announcements"

    id: Mapped[int] = mapped_column(primary_key=True)
    topic: Mapped[str] = mapped_column(String(32), index=True)  # jlpt | mba | custom
    title: Mapped[str] = mapped_column(String(512))
    summary: Mapped[str] = mapped_column(Text, default="")
    source_url: Mapped[str] = mapped_column(String(1024))
    source_domain: Mapped[str] = mapped_column(String(255), index=True)
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    first_seen_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), index=True
    )
    # hash nội dung chuẩn hoá — dedup, phát hiện mục MỚI
    content_hash: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    verified: Mapped[bool] = mapped_column(Boolean, default=False)
    score: Mapped[float] = mapped_column(Float, default=0.0)  # độ khớp chủ đề 0..1
    # ngày regex tự phát hiện trong tin (JSON list [{date, label, raw}]) — GỢI Ý,
    # chưa kiểm chứng. Lịch chuẩn nằm ở exam_events.
    extracted_dates: Mapped[str | None] = mapped_column(Text, nullable=True)
